### ISSUE TITLE
Problems with new gray dashed focus style

### ISSUE THREAD

**Issue by maya** [uswds/uswds#2142]
Hi all, I have some concerns about changing focus from vibrant blue to light gray: https://github.com/18F/web-design-standards/pull/2105

I agree that the current blue halo is problematic, but I design *a lot* on gray backgrounds and this will be an issue to have visible focus for many of my use cases. I think a more resilient approach is something like GOV.UK's solid gold outline and solid gold background on text links. This is also used on USA.gov (orange) /cc @davidstenger. While it's always hard to find the right focus color bc of different background colors, not many designs are bright orange/yellow, so it shouldn't conflict often. The dashes are also visually busy (personal opinion). Using a color with vibrance will help its visibility.

Was this approach tested with low-vision end-users? Is there research pointing to this approach?

Just some thoughts and would love to hear about the approach and what others think.

**Comment by rtwell** [comment]
see my response here: https://github.com/18F/web-design-standards/pull/2112#issuecomment-327863317

also, correct, there is no color that will work on every website. gray + dotted is the most neutral, and is less likely to be confused with interactive elements (the shadow focus styling looking like a simple shadow, not focus). and while, the gov.uk solution is nice, it is not one focus for all elements, but requires different code for different elements (as i understand it, anyway). also colors like orange, red, yellow and green mean different things in forms, which the WDS are used frequently. when there is an error or a warning and then that element is focused, it's much harder to track.

**Comment by maya** [comment]
Thanks for the explanation, I understand how you came to that decision with those constraints now. 

If the solid background like GOV.UK introduces too much complexity, we can stick with simple outline. I don't recall seeing gold used on the [forms recommendations](https://standards.usa.gov/components/form-controls/#text-input), but assuming that it could come later. Also, from what I gather, adding space around certain elements requires extra code too.

Perhaps testing with people with disabilities and developers could help stress test this solution. I'm generally cautious of making decisions for PWDs and just want to ensure the best solution come out for them (as well as minimizing introducing new problems for developers).

**Comment by asarazor** [comment]
I second that this change requires some testing to back up its efficacy. It seems the change has been made based on design gut feel and developer convenience. Those are valid in themselves, but for an accessibility feature there seems to be a distinct lack of data here.

**Comment by davidstenger** [comment]
Both Section 508 and WCAG 2.0 state that there must be a (“well defined”) visual indication of focus, but neither explain what is considered sufficient. So it could be argued that relying on the native browser focus indication is sufficient for 508 compliance and that it would pass a 508 audit (although it probably depends on who the auditor is and what he or she feels is sufficient). With this in mind, the first requirement in your style guide would, of course, be that there is a visual indication of focus, to begin with. There are a few things to keep in mind here:

- Some CSS developers like to disable the native focus indicator. For example, they will define “outline: 0;” in CSS (hiding the dotted line), or use scripting to immediately move focus away from a focused element so that you’ll never see the highlight (making the focusable control keyboard inaccessible at the same time). Such techniques should never be allowed and always be flagged as crucial accessibility obstacles.

- In older versions of IE, only natively focusable elements (e.g. links and form controls) show the native indicator. If other elements (such as <divs> and <spans> used for ARIA widgets) are made focusable using the tabindex attribute, those elements will be able to receive programmatic focus, but will not show the native focus indicator.

- Regardless of how the focus is visualized, it must always be exposed programmatically. The object inspection tools are used to test for this.

Having said all this, I personally find the default dotted line in IE and the light blue line used in Chrome, Safari and FFX not sufficient enough in most cases. It’s often difficult to locate focus visually (especially on complex pages), and generally the dotted line doesn’t look very nice either. I usually recommend using CSS to provide a more clearly defined focus outline of at least 2px. That way it’s always clear for keyboard users (even those who don’t have perfect vision) where the current focus is. For example, uses the following styles for focus:

_a:focus, input:focus, div[role=group]:focus, button:focus, testarea-focus,button:focus, select:focus {
   outline: 2px solid #24889C;
   outline-offset: 2 px;
}_

The advantage of the outline property is that it doesn’t affect the positioning of the element like the border style does. Of course, depending on the content in can be useful to provide additional visual cues to visualize focus (e.g. CSS based icons, changes in background and foreground colors, font-weight, text decoration, and so on).

Of course, there are also a few issues with the snippet above:

- Visual designers and developers generally don’t like such a strongly defined focus indicator. If they oppose it, a fallback solution could be to provide this focus style as an alternative stylesheet that users can opt in to.

- Internet Explorer: Outline styles are not supported by older browsers, in particular, IE7 and below. The same goes for the :focus pseudo selector (although for links this can be resolved by simultaneously using the :active pseudo selector). If you want a custom focus style to be visible in IE7 as well, then you’ll have to use scripting to add a focus event handler to all focusable elements, in which a class name is added to the element (and similarly add a blur event handler that removes this class name). Since you can’t depend on the outline property in CSS, you would have to visually indicate the focus using a different way (e.g. changing the border, adding underlines, etc.). In other words, it’s a lot of hassle and hacking.

So to summarize: Relying on the native visual focus indication could be considered as 508 compliant (it would be debatable for some though), but for true accessibility, it’s better to provide a more clearly defined indication using CSS as part of your style guide. It’s not something I think you can really enforce teams to do though, it’s a recommendation that a team can follow if they agree with it. I would recommend sticking to the modern approach (using :focus and “outline”). Whatever approach is chosen, it’s always important that the visual indication of focus is consistent throughout the site. So that means it’s best to use the same focus highlight for all focusable content, regardless of whether it’s text links, image links, form controls, or ARIA widgets.

**Comment by hbillings** [comment]
I've run into some issues with this also -- here are a couple of screenshots:

The contrast with the light colors in the project I was just on was pretty rough. Also notice how it looks odd with the search button:
<img width="736" alt="screen shot 2017-11-07 at 1 29 08 pm" src="https://user-images.githubusercontent.com/509309/32516252-f67f6422-c3bf-11e7-8dda-9a58bd539761.png">

The offset on it can also be problematic for some elements, like these tabs:
<img width="427" alt="screen shot 2017-11-07 at 1 30 19 pm" src="https://user-images.githubusercontent.com/509309/32516238-e68dfe0c-c3bf-11e7-9149-4c188db1c91c.png">

I'd prefer a bright contrasting color instead of something that recedes (that said, this *does* work great on white backgrounds with lots of form fields -- maybe it's worth considering multiple focus styles).

**Comment by thisisdano** [comment]
I feel like a total broken record always looking to gov.uk, but I _do_ like the gov.uk block highlight style (as distinct from their text highlight style)

highlighted:
<img width="667" alt="screen shot 2017-11-07 at 1 01 13 pm" src="https://user-images.githubusercontent.com/11464021/32517607-c60de6c8-c3bb-11e7-977d-3685f8f10737.png">

not highlighted:
<img width="661" alt="screen shot 2017-11-07 at 1 01 07 pm" src="https://user-images.githubusercontent.com/11464021/32517609-c8b7b534-c3bb-11e7-8268-e72e8ef8b6dd.png">

As you can see, there's some overlap with adjacent elements, but `outline` is probably the right call here, as it doesn't affect layout as `border` might.

`$highlight-color` could also be a var just in case you're designing a yellow-heavy site 😬

**Comment by thisisdano** [comment]
I'd prefer to have variance in color rather than style, and I prefer a solid outline to dots.

**Comment by maya** [comment]
We can use any contrast ratio and still be 508 compliant since it uses WCAG 2.0 which doesn’t have a contrast requirement.

However, WCAG 2.1 **does** have a contrast requirement for focus states ([1.4.11 Non-text Contrast](https://www.w3.org/TR/WCAG21/#non-text-contrast)) which is at least 3:1 (AA for large text). Section 508 doesn’t have any foreseeable plans to require 2.1 as they only just incorporated 2.0 this year.

If we are OK with aligning with WCAG 2.0, we can use the current choice of Gold 30v (2.22 contrast). If we want to be future friendly and some may argue more accessible, we should align with WCAG 2.1. This would be the 40 values of the colors (although I found those colors didn’t always meet AA large text)

I recommend staying away from red and green bc they are used as error and success states on forms.

I tested yellows, oranges, and blues. Did not test teal and purple, but people could potentially use it if it works with their color scheme.

## Color options AA
#### Red warm 30v `#e67549`
3 on white
7 on black
<img width="116" alt="red warm 30v e67549" src="https://user-images.githubusercontent.com/5249443/42862969-61d69908-8a16-11e8-9eb5-8c825d2e4953.png">

#### Orange 40 `#eb8028`
2.74 on white ❌ 
7.65 on black
Passes AA if you change text color to `#E87516` New contrast ratio: 3.01
<img width="117" alt="orange 40 eb8028" src="https://user-images.githubusercontent.com/5249443/42862975-66bbb214-8a16-11e8-991a-d4d82a505035.png">

#### Gold 40 `#ad8b65`
3.15 on white
6.66 on black
<img width="115" alt="gold 40 ad8b65" src="https://user-images.githubusercontent.com/5249443/42862979-6ae5dde2-8a16-11e8-97da-50554697f1c5.png">

#### Gold 40v `#c98a0c`
2.95 on white ❌
7.12 on black
Passes AA if you change text color to `#C7880C` New contrast ratio: 3.02
<img width="115" alt="gold 40v c98a0c" src="https://user-images.githubusercontent.com/5249443/42862984-6ed0f540-8a16-11e8-8b68-ba7069d2499a.png">

#### Yellow 40 `#ab9149`
3.05 on white
6.88 on black
<img width="115" alt="yellow 40 ab9149" src="https://user-images.githubusercontent.com/5249443/42862990-7386a940-8a16-11e8-9b1d-716ff5d1a3eb.png">

#### Blue 40v `#009afa`
3 on white
7 on black
<img width="117" alt="blue 40v 009afa" src="https://user-images.githubusercontent.com/5249443/42862998-7ad3e26c-8a16-11e8-9365-8649440fb51a.png">

#### Indigo cool 40 `#6d91ed`
3.03 on white
6.92 on black
<img width="116" alt="indigo cool 40 6d91ed" src="https://user-images.githubusercontent.com/5249443/42863005-7fd9c47a-8a16-11e8-989d-8d88be5cdca1.png">

## Current
#### Gold 30v `#e6a100`
2.22 on white ❌
9.46 on black
Passes AA if you change text color to `#C48900` New contrast ratio: 3.03
<img width="118" alt="gold 30v" src="https://user-images.githubusercontent.com/5249443/42863031-988b252c-8a16-11e8-953d-2c445d4b7bbd.png">

---

### Why yellow? 
Yellow is the most visible color on the spectrum, less likely to conflict with most color schemes, triad to default blue
### Why blue? 
Light blue is used as browser's default focus color so people are accustomed to it and trust it, using a monochromatic / lighter tone of our default blue makes it cohesive, potential to have issues with a blue scheme with a bright/light background but this may not be often
### Why orange? 
A mix of red and yellow it has high visibility, orange is complementary to default blue, the negative is it can be similar to the red error state especially if it doesn’t have enough yellow.

**Comment by jenniferthibault** [comment]
Hey @maya, what do you think about adding some more data here in order to make a good recommendation? Seeing these against white doesn't capture the different background color use cases you know about, and mentioned above. 

Here's one way you might be able to do that: 
What are the most common ~3-5 solid background colors used on sites using USWDS? When you get that list, then add mockups of these colors against those backgrounds and see if any options disqualify themselves.

**Comment by maya** [comment]
It's also interesting to see the focus that the WCAG website uses: thicker gray underline and light gray background. The blue halo is the browser default.
<img width="804" alt="screen shot 2018-07-19 at 12 07 22 pm" src="https://user-images.githubusercontent.com/5249443/42964698-acc87c56-8b4c-11e8-86ae-218f4e6100fc.png">

**Comment by maya** [comment]
I met with the 18F visual design team today to discuss this issue. 

The team recomended we meet WCAG 2.1's requirements of AA large text (3:1 contrast ratio) in order to be future friendly and the most accessible.

The team favored **Blue 40v** `#009afa` as the focus state. They recommend adding a focus offset only to buttons (not links) to make sure it's visible on blue buttons.

#### Blue 40v `#009afa`
3 on white
7 on black
<img width="116" alt="screen shot 2018-07-19 at 3 13 55 pm" src="https://user-images.githubusercontent.com/5249443/42973021-68fe1c3c-8b66-11e8-9e04-cf1c529da4c1.png">
One designer mentioned the original focus being too thick so I changed it to 3px thickness and 3px offset.

### Blue focus on landing page template
![Blue focus](http://g.recordit.co/qeORP8gwce.gif)

**Comment by thisisdano** [comment]
Since `3px` isn't in the new spacing unit palette, could we try it at `2px` and `4px` as well?

**Comment by maya** [comment]
@thisisdano here is what it looks like in 2, 3, 4:

2px
<img width="161" alt="screen shot 2018-07-19 at 3 40 53 pm" src="https://user-images.githubusercontent.com/5249443/42973978-4e8f25d6-8b6a-11e8-889e-a7d955fd04f1.png">

3px
<img width="160" alt="screen shot 2018-07-19 at 3 41 44 pm" src="https://user-images.githubusercontent.com/5249443/42973980-4ec56c04-8b6a-11e8-93a6-de1e4f7c9937.png">

4px
<img width="161" alt="screen shot 2018-07-19 at 3 41 01 pm" src="https://user-images.githubusercontent.com/5249443/42973979-4eaa5004-8b6a-11e8-83ae-f648b3732746.png">

**Comment by thisisdano** [comment]
I would probably move in the direction of 4px — better to have one that has too much presence than too little, I'd say

**Comment by maya** [comment]
Did you mean 4px outline and 4px offset?

4px outline and 4px offset
<img width="169" alt="screen shot 2018-07-19 at 3 56 37 pm" src="https://user-images.githubusercontent.com/5249443/42974534-8a9ba1ec-8b6c-11e8-8880-338096166b18.png">

4px outline and 2px offset
<img width="165" alt="screen shot 2018-07-19 at 3 56 48 pm" src="https://user-images.githubusercontent.com/5249443/42974535-8d2e2c18-8b6c-11e8-95d8-33f01ed26cc0.png">

Original: 3px outline and 3px offset
<img width="163" alt="screen shot 2018-07-19 at 3 59 40 pm" src="https://user-images.githubusercontent.com/5249443/42974594-c7de63dc-8b6c-11e8-880d-0b84e5a59372.png">

/cc @jenniferthibault @AvivaOskow @austinhernandez

**Comment by maya** [comment]
Should also add orange was a solid second place to blue.

🥇Blue 40v `#009afa`  
>I would pick blue in this instance — it feels like it belongs in the system, looks nicest, and to me, feels in -line with my expectations which are set by that fuzzy blue browser default style.

>IMO blue looks the nicest but you can’t see it on the buttons very well [@maya's note: before the added offset]

🥈Orange 40 `#eb8028`  
>bright orange could look like an error state but it seems the most obvious to me!

>The orange is most vibrant against the blue background, but feels slightly out of place/surprising against white (but maybe I’d get used to it)

🥉Gold 40v `#c98a0c`  
>the gold is definitely out for me here though, too muddled on white

**Comment by thisisdano** [comment]
hmmmmmm. This whole issue has me a bit twisted, as I think the WCAG 2.1 guidance is not as helpful as I would have expected. That is to say, it is practically impossible to follow, understanding that focus states occur over all kinds of background conditions and it is not context aware. I also don't really think that AA Large is a good solution for focus state either (whatever that means, AA Large against the background? against the text? is a small, but high contrast indicator preferable to a larger, less contrast indicator? who knows!?)  — I think it begins to get the color too close to common UI colors (in terms of level) and starts to introduce ambiguities, like the necessity of offsetting the outline around buttons since you have to use a color that looks like a button color! If your vision was blurry, even the offset wouldn't be enough the fix that ambiguity.

Yuck.

I just think the offset looks pretty janky no matter what it is, and even `outline` itself, while a good solution, looks pretty crude. Would `box-shadow: 0 0 0 4px [color]` be an acceptable solution (as we see on some focus states on Github)? At least then it follows the rounded edge of buttons.

And I think I don't really like 40 level focus indicators. And it's not just "liking" — AA Large contrast doesn't seem like a thoughtful accessibility solution to this problem, more of a bolt-on. Am I nuts for thinking this? Focus state is about contrast, but also about highlighting and context — and almost _anything_ would be an improvement on the default, which is fuzzy, low contrast, and gray. I really think that a 20v or 30v level indicator (like gold-20c or orange 30v), if sufficiently vivid, would be a better solution.

Anyhow, sorry for being so prickly on this topic. It's important, but I don't think there's reasonable best-practice guidance yet — it seems like even the experts are struggling.

**Comment by maya** [comment]
Success Criterion 1.4.11 Non-text Contrast states its the contrast ratio against adjacent colors. This means the background color for the most part.

We can use box-shadow but that would be we can't offset it and get the background color to come through like we do now, we'd have to use multiple borders to do this and choose a border color (ie white) to fake it and then you'd have a strip of white and blue on anything besides a white background.

I don't know what we like is as important as what works best for the folks who rely on focus to navigate the web. Maybe we should ask them?

The options above are not aiming to reach 3:1 on any background, just white and black. And something that is visible on most cases i.e. doesn't disappear like the gray did.

**Comment by thisisdano** [comment]
That makes me a little interested in a potential combinations of colors, like multiple drop shadows, to account for changing background color context. I don't necessarily think offset is a crucial part of the solution here. It could be, but not necessarily. 

It would also be fascinating and useful to have real feedback and information from actual people who rely on focus-tabbing for navigation.

**Comment by thisisdano** [comment]
Here's potentially an interesting solution:

```
a:focus {
  box-shadow: 0 0 1px rgba(0, 0, 0 ,.9), 0 0 5px #FAAF00;
}
```

This uses a thin, dark stroke and a larger gold-30v highlight to result in something that should have AA Large contrast against any background, has good vividness, and provides distinction against filled buttons.

<img width="534" alt="screen shot 2018-07-20 at 10 16 25 am" src="https://user-images.githubusercontent.com/11464021/43016123-ac9c8a50-8c06-11e8-82ae-2e533411877f.png">
<img width="418" alt="screen shot 2018-07-20 at 10 16 36 am" src="https://user-images.githubusercontent.com/11464021/43016128-af0e0b10-8c06-11e8-9439-4334eb924bee.png">
<img width="560" alt="screen shot 2018-07-20 at 10 16 47 am" src="https://user-images.githubusercontent.com/11464021/43016130-b13da760-8c06-11e8-8a4e-1a7c746e081b.png">

**Comment by maya** [comment]
Many thanks for the feedback, everyone! The blue focus is implemented in #2604 and can be seen in use [here](https://federalist-proxy.app.cloud.gov/preview/uswds/uswds/release-2.0/components/preview/layout--landing.html). 

We have some other strong options to consider like orange and @thisisdano's suggestion. We'd also like to test these options with folks that rely on focus to navigate the web, and maybe some have vision impairments as well, so that we can recommend the option that works best.
